### PR TITLE
add remove metadata

### PR DESCRIPTION
**Pull request by yasirroni**
Add new argument for clean and filter "-M" or "--remove-notebook-metadata" to "remove notebook metadata.

```python
    clean_parser.add_argument(
        "-M",
        "--remove-notebook-metadata",
        action="store_true",
        help="remove notebook metadata",
    )
```

re-PR from https://github.com/srstevenson/nb-clean/pull/163

### CODE REVIEW

**Comment by yasirroni** [comment]
Fails due to pylint:

```shell
poetry run pylint src  
************* Module nb_clean
src\nb_clean\__init__.py:147:0: R0913: Too many arguments (6/5) (too-many-arguments)

-----------------------------------
Your code has been rated at 9.95/10
```

**Comment by yasirroni** [comment]
Maybe refactor to keyword-only-argument? See [SO answer](https://stackoverflow.com/a/68744457/11671779) and [PEP 3102](https://peps.python.org/pep-3102/).

**Comment by yasirroni** [comment]
Any update, @srstevenson 

The fails is not mandatory and can be fixed by making all args as keyword-only-argument

**Comment by srstevenson** [comment]
I've held off from merging this because I'm not yet convinced there's a real user requirement for it. Most of the notebook metadata doesn't change frequently, and so doesn't contribute to the noise in VCS diffs which nb-clean was written to address.

**Comment by yasirroni** [comment]
Different editor such as Google Colab vs VSCode and different Python version across user produce different metadata. That is the use case.

Not convinced enough?

File: README.md [hunk 1]
```diff
@@ -119,6 +119,13 @@ To preserve cell outputs, use:
 nb-clean add-filter --preserve-cell-outputs
 ```
 
+By default, `nb-clean` did not clean notebook metadata. To clean notebook
+metadata use:
```
Comment (haplav) [inline README.md:123]:
```suggestion
To clean notebook metadata, use:
```

File: src/nb_clean/__init__.py [hunk 2]
```diff
@@ -232,6 +245,8 @@ def clean_notebook(
         The notebook.
     remove_empty_cells : bool, default False
         If True, remove empty cells.
+    remove_notebook_metadata : bool, default False
+        If True, remove notebook metada.
```
Comment (haplav) [inline src/nb_clean/__init__.py:249]:
```suggestion
        If True, remove notebook metadata.
```

**Review by haplav — COMMENTED** [review]
Found some typos

File: src/nb_clean/cli.py [hunk 3]
```diff
@@ -145,6 +145,7 @@ def clean(args: argparse.Namespace) -> None:
         notebook = nb_clean.clean_notebook(
             notebook,
             remove_empty_cells=args.remove_empty_cells,
+            remove_notebook_metadata=args.remove_notebook_metadata,
```
Comment (haplav) [inline src/nb_clean/cli.py:148]: I guess you're supposed to do the same for `check()` and `add_filter()`, and update the docstrings of all three functions.

**Comment by yasirroni** [comment]
@haplav 

>  "metadata": {
>  "language_info": {
>   "name": "python"
>  },

Since notebook metadata language_info is indeed needed, current PR remove this metadata. How about add other parameter such as --add-language-metadata NAME, with default name to Python?

**Comment by yasirroni** [comment]
Another approach is to change this PR into `--preserve-metadata language_info metadata2 metadata3`. But, support of hierarchy of metadata sometimes are needed.

But, based on your example, the `language_info` is destroyed and replaced with `codemirror_mode`. That is another problem. And that is not what happen on my side. Here is my case:
Original:
```
 "metadata": {
  "language_info": {
   "name": "python"
  }
```
After VSCode open:
```
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.10 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.8.10"
  },
  "vscode": {
   "interpreter": {
    "hash": "..."
   }
  }
```

**Comment by haplav** [comment]
> 
> > "metadata": {
> > "language_info": {
> > "name": "python"
> > },
> 
> Since notebook metadata language_info is indeed needed, current PR remove this metadata. How about add other parameter such as --add-language-metadata NAME, with default name to Python?

I don't understand @yasirroni, I don't mind having this in the cleaned notebook with hard-wired `"name": "python"`. It seems rather unlikely this would need to be changed.

**Comment by haplav** [comment]
> Another approach is to change this PR into `--preserve-metadata language_info metadata2 metadata3`. But, support of hierarchy of metadata sometimes are needed.
> 

I don't think we need to complicate things right now.

> But, based on your example, the `language_info` is destroyed and replaced with `codemirror_mode`. That is another problem. And that is not what happen on my side. Here is my case: Original:
> 

Ah OK, I think you read it wrong; `language_info` is NOT destroyed; `codemirror_mode` gets just ADDED to it.

File: src/nb_clean/__init__.py [hunk 4]
```diff
@@ -159,6 +164,8 @@ def check_notebook(
         The notebook.
     remove_empty_cells : bool, default False
         If True, also check for the presence of empty cells.
+    remove_notebook_metadata : bool, default False
+        If True, also check for the presence of nonempty notebok metadata.
```
Comment (haplav) [inline src/nb_clean/__init__.py:168]:
```suggestion
        If True, also check for the presence of nonempty notebook metadata.
```

**Comment by haplav** [comment]
Please let GitHub apply suggestions and resolve discussions automatically as described in https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request

File: README.md [hunk 5]
```diff
@@ -119,6 +119,13 @@ To preserve cell outputs, use:
 nb-clean add-filter --preserve-cell-outputs
 ```
 
+By default, `nb-clean` did not clean notebook metadata. To clean notebook
+metadata, use:
```
Comment (haplav) [inline README.md:123]:
```suggestion
To clean notebook metadata, use:
```

**Comment by yasirroni** [comment]
> Please let GitHub apply suggestions and resolve discussions automatically as described in https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request

Did not know that feature exist. Sorry, Sir.

**Comment by yasirroni** [comment]
> Ah OK, I think you read it wrong; `language_info` is NOT destroyed; `codemirror_mode` gets just ADDED to it.

Yeah, but the structure changes, since `name` no longer the child of `language_info`.

> > > "metadata": {
> > > "language_info": {
> > > "name": "python"
> > > },
> > 
> > 
> > Since notebook metadata language_info is indeed needed, current PR remove this metadata. How about add other parameter such as --add-language-metadata NAME, with default name to Python?
> 
> I don't understand @yasirroni, I don't mind having this in the cleaned notebook with hard-wired `"name": "python"`. It seems rather unlikely this would need to be changed.

But, I also works in Octave and octave-based notebook did not use that metadata. Here is a new notebook in Octave created using jupyterlab:

```json
 "metadata": {
  "kernelspec": {
   "display_name": "Octave",
   "language": "octave",
   "name": "octave"
  }
 },
```

**Comment by yasirroni** [comment]
I think, two options might be good.

Hard clean: like this PR.
Soft clean: like this PR, but skip for

```json
"metadata": {
  "kernelspec": {
   "language": "..."
  },
  "language_info": {
   "name": "...",
  },
 },
```

**Comment by haplav** [comment]
> Yeah, but the structure changes, since `name` no longer the child of `language_info`.

No. Please, take a good look - `name` is contained in `kernelspec` and `codemirror_mode` but also still directly in `language_info`:

```json
 "metadata": {
  "kernelspec": {
   ...
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "python",
    ...
   },
   ...
   "name": "python",
   ...
  },
  ...
 },
```
 
> I think, two options might be good.

Once again, I'm fine with keeping just `metadata/language_info/name` (if already present). It seems to be the same in all cases you and I mentioned. But it doesn't make any sense to preserve `kernelspec` exactly because

> But, I also works in Octave and octave-based notebook did not use that metadata. Here is a new notebook in Octave created using jupyterlab:
> 
> ```
>  "metadata": {
>   "kernelspec": {
>    "display_name": "Octave",
>    "language": "octave",
>    "name": "octave"
>   }
>  },
> ```

It's changed based on what interpreter the kernel uses. Keeping just empty `metadata`, as this PR already does, looks also fine to me. I don't think any interpreter would depend on having anything in there.

**Comment by yasirroni** [comment]
> 

@haplav

**Comment by yasirroni** [comment]
> No. Please, take a good look - name is contained in kernelspec and codemirror_mode but also still directly in language_info:

@haplav oops. i'm sorry. It will be easier than.
_
Okay, maybe just let this PR as it is and later add feature to skip language_info

**Comment by yasirroni** [comment]
Any chance this can be re-opened? I'm willing to fix merge conflict. @srstevenson

**Pull request closed without merging**